File: polyglot/leaderboard_sort.r
#!/usr/bin/env Rscript

# Leaderboard sorting (highest balance first).
# Input format: lines like "username creditsMinor"

lines <- readLines(file("stdin"), warn = FALSE)
lines <- lines[nzchar(trimws(lines))]

if (length(lines) == 0) {
  quit(status = 0)
}

parts <- strsplit(lines, "\\s+")
username <- sapply(parts, function(x) x[[1]])
credits <- as.integer(sapply(parts, function(x) x[[2]]))

df <- data.frame(username = username, creditsMinor = credits, stringsAsFactors = FALSE)
df <- df[order(-df$creditsMinor), ]

for (i in seq_len(nrow(df))) {
  cat(i, df$username[i], df$creditsMinor[i], "\n")
}

File: polyglot/leaderboard_sort.hs
#!/usr/bin/env runghc

-- Leaderboard sorting (highest balance first).
-- Input format: lines like "username creditsMinor"

import Data.List (sortOn)
import Data.Ord (Down(..))

parseLine :: String -> Maybe (String, Int)
parseLine s =
  case words s of
    (u:c:_) -> Just (u, read c)
    _       -> Nothing

main :: IO ()
main = do
  input <- getContents
  let rows = [ r | Just r <- map parseLine (lines input) ]
      sorted = sortOn (Down . snd) rows
  mapM_ putStrLn [ show i ++ " " ++ u ++ " " ++ show c | (i,(u,c)) <- zip [1..] sorted ]

